### PR TITLE
Add a popup window to exit full screen mode.

### PR DESCRIPTION
**Pull request by joone**
There was no way to exit full screen mode in Windows so this CL adds a pop-up window to allow users to exit full screen mode by clicking the message to "Exit Full Screen". The pop-up window is automatically disappeared, but it can be reactivated by placing a mouse at the top of the screen in the center of the window.

BUG=XWALK-4861

### CODE REVIEW

File: runtime/browser/ui/native_app_window_views.cc [hunk 1]
```diff
@@ -144,6 +147,16 @@ void NativeAppWindowViews::SetFullscreen(bool fullscreen) {
       xwalk::NOTIFICATION_FULLSCREEN_CHANGED,
       content::Source<NativeAppWindow>(this),
       content::NotificationService::NoDetails());
+
+  // Add a popup window to exit full screen mode.
+  if (fullscreen) {
+    gfx::NativeWindow parent = web_contents_->GetTopLevelNativeWindow();
+    exclusive_access_bubble_.reset(
+        new ExclusiveAccessBubbleViews(this, GURL(), parent));
+    exclusive_access_bubble_.get()->UpdateContent(GURL());
```
Comment (pozdnyakov) [inline runtime/browser/ui/native_app_window_views.cc]: nit: get() is not needed

File: runtime/browser/ui/desktop/exclusive_access_bubble.cc [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+bool ExclusiveAccessBubble::IsWatchingMouse() const {
+  return mouse_position_checker_.IsRunning();
+}
+
+void ExclusiveAccessBubble::CheckMousePosition() {
+  // Desired behavior (without the "simplified" flag):
+  //
+  // +------------+-----------------------------+------------+
+  // | _  _  _  _ | Exit full screen mode (F11) | _  _  _  _ |  Slide-in region
```
Comment (pozdnyakov) [inline runtime/browser/ui/desktop/exclusive_access_bubble.cc]: I think we'd rather need "close" instead of "Exit full screen mode", for the fullscreen apps it should not be possible to switch to windowed mode. @darktears WDYT?

Comment (joone) [inline runtime/browser/ui/desktop/exclusive_access_bubble.cc]: I agree with you. I need "close" button.

Comment (joone) [inline runtime/browser/ui/desktop/exclusive_access_bubble.cc]: In this case, we need two UIs for full screen apps and web pages that use the full screen API.

Comment (darktears) [inline runtime/browser/ui/desktop/exclusive_access_bubble.cc]: I agree too

File: runtime/browser/ui/native_app_window_views.cc [hunk 3]
```diff
@@ -144,6 +147,24 @@ void NativeAppWindowViews::SetFullscreen(bool fullscreen) {
       xwalk::NOTIFICATION_FULLSCREEN_CHANGED,
       content::Source<NativeAppWindow>(this),
       content::NotificationService::NoDetails());
+
+  ExclusiveAccessBubbleType bubble_type =
```
Comment (pozdnyakov) [inline runtime/browser/ui/native_app_window_views.cc]: `ExclusiveAccessBubbleType bubble_type  = (create_params_.state == ui::SHOW_STATE_FULLSCREEN) ?                                                                   EXCLUSIVE_ACCESS_BUBBLE_TYPE_APPLICATION_FULLSCREEN_CLOSE_INSTRUCTION : EXCLUSIVE_ACCESS_BUBBLE_TYPE_BROWSER_FULLSCREEN_EXIT_INSTRUCTION;`

File: runtime/browser/ui/native_app_window_views.cc [hunk 4]
```diff
@@ -144,6 +147,24 @@ void NativeAppWindowViews::SetFullscreen(bool fullscreen) {
...
+    EXCLUSIVE_ACCESS_BUBBLE_TYPE_BROWSER_FULLSCREEN_EXIT_INSTRUCTION;
+
+  if (create_params_.state == ui::SHOW_STATE_FULLSCREEN) {
+    bubble_type =
+      EXCLUSIVE_ACCESS_BUBBLE_TYPE_APPLICATION_FULLSCREEN_CLOSE_INSTRUCTION;
+  }
+
+  // Add a popup window to exit full screen mode or end the application.
+  if (fullscreen) {
```
Comment (pozdnyakov) [inline runtime/browser/ui/native_app_window_views.cc]: `bubble_type` should be created here

File: runtime/browser/ui/desktop/exclusive_access_bubble.cc [hunk 5]
```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2012 The Chromium Authors. All rights reserved.
```
Comment (pozdnyakov) [inline runtime/browser/ui/desktop/exclusive_access_bubble.cc]: once the original code is modified, should crosswalk copyright be also present?

File: runtime/browser/ui/desktop/exclusive_access_bubble.h [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+#include "ui/gfx/animation/animation_delegate.h"
+#include "ui/gfx/geometry/point.h"
+#include "url/gurl.h"
+
+namespace gfx {
+class Rect;
+}
+
+// Describes the contents of the fullscreen exit bubble.
+enum ExclusiveAccessBubbleType {
```
Comment (pozdnyakov) [inline runtime/browser/ui/desktop/exclusive_access_bubble.h]: I'd move it inside `ExclusiveAccessBubble` class, then it could be:
```
enum Type {
TYPE_NONE = 0,
TYPE_BROWSER_FULLSCREEN_EXIT_INSTRUCTION,
TYPE_APPLICATION_FULLSCREEN_CLOSE_INSTRUCTION
};
```

File: runtime/browser/ui/desktop/exclusive_access_bubble.h [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+//
+// Notification display design note: if the #simplified-fullscreen-ui flag is
+// enabled, the bubble has the following behaviour:
+// - The bubble is shown for kInitialDelayMs, then hides.
+// - After a bubble has been shown, notifications are suppressed for
+//   kSnoozeNotificationsTimeMs, to avoid bothering the user. After this time
+//   has elapsed, the next user input re-displays the bubble.
+class ExclusiveAccessBubble : public gfx::AnimationDelegate {
+ public:
+  explicit ExclusiveAccessBubble(const GURL& url,
```
Comment (pozdnyakov) [inline runtime/browser/ui/desktop/exclusive_access_bubble.h]: `explicit ` is meaningless since we have more than 1 arg here.
Do we actually need `url` arg? how is it used?

File: runtime/browser/ui/desktop/exclusive_access_bubble_views.cc [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+}
+
+void ExclusiveAccessBubbleViews::ExclusiveAccessView::LinkClicked(
+    views::Link* link,
+    int event_flags) {
+  bubble_->ExitExclusiveAccess();
+}
+
+void ExclusiveAccessBubbleViews::ExclusiveAccessView::UpdateContent(
+    const GURL& url,
```
Comment (pozdnyakov) [inline runtime/browser/ui/desktop/exclusive_access_bubble_views.cc]: `url` is unused, right?

File: runtime/browser/ui/desktop/exclusive_access_bubble_views.cc [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    ExclusiveAccessBubbleType bubble_type) {
+
+  DCHECK(message_label_);
+  message_label_->SetText(bubble_->GetCurrentMessageText());
+
+  base::string16 link_text;
+
+  if (bubble_type ==
```
Comment (pozdnyakov) [inline runtime/browser/ui/desktop/exclusive_access_bubble_views.cc]: nit: `switch (bubble_type)` would be better

File: runtime/browser/ui/desktop/exclusive_access_bubble_views.cc [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+void ExclusiveAccessBubbleViews::OnWidgetVisibilityChanged(
+    views::Widget* widget,
+    bool visible) {
+  UpdateMouseWatcher();
+}
+
+void ExclusiveAccessBubbleViews::ExitExclusiveAccess() {
+  if (bubble_type_ ==
+      EXCLUSIVE_ACCESS_BUBBLE_TYPE_APPLICATION_FULLSCREEN_CLOSE_INSTRUCTION)
+    bubble_view_context_->GetNativeAppViews()->Close();
```
Comment (pozdnyakov) [inline runtime/browser/ui/desktop/exclusive_access_bubble_views.cc]: if there is a one more pending window (e.g. a  popup) the application will not exit. I think it's better if this user action is propagated to `Runtime` (and then to `Application` so that it is closed) through `NativeAppWindowDelegate`

Comment (joone) [inline runtime/browser/ui/desktop/exclusive_access_bubble_views.cc]: I tried to call delegate_->OnWindowDestroyed(), but crosswalk becomes freeze if there are multiple windows.

File: runtime/browser/ui/desktop/exclusive_access_bubble.h (hunk 7 above)
Comment (joone) [inline runtime/browser/ui/desktop/exclusive_access_bubble.h]: no, I just left the url for the future use so I removed it.

File: runtime/browser/ui/desktop/exclusive_access_bubble_views.cc (hunk 10 above)
Comment (joone) [inline runtime/browser/ui/desktop/exclusive_access_bubble_views.cc]: I fixed the problem.

**Comment by joone** [comment]
@pozdnyakov PTAL

File: extensions/extension_process/xwalk_extension_process.cc [hunk 11]
```diff
@@ -124,7 +124,7 @@ bool XWalkExtensionProcess::CheckAPIAccessControl(
   PermissionCacheType::iterator iter =
       permission_cache_.find(extension_name + api_name);
   if (iter != permission_cache_.end())
-    return iter->second;
+    return iter->second != ALLOW_ONCE;
```
Comment (pozdnyakov) [inline extensions/extension_process/xwalk_extension_process.cc]: this and other extensions changes look unrelated

File: runtime/browser/ui/native_app_window_views.cc [hunk 12]
```diff
@@ -158,6 +180,15 @@ void NativeAppWindowViews::Close() {
   window_->Close();
 }
 
+void NativeAppWindowViews::ExitApplication() {
+  auto app_service =
```
Comment (pozdnyakov) [inline runtime/browser/ui/native_app_window_views.cc]: it should not be like this.. pls add` NativeAppWindowDelegate::OnApplicationExitRequested` (or invent a better name) and propagate it further

File: extensions/extension_process/xwalk_extension_process.cc (hunk 11 above)
Comment (joone) [inline extensions/extension_process/xwalk_extension_process.cc]: This change is from https://github.com/crosswalk-project/crosswalk/commit/231ec374c53fed0166357aa4e013c8b246abaddd

File: runtime/browser/ui/native_app_window_views.cc (hunk 12 above)
Comment (joone) [inline runtime/browser/ui/native_app_window_views.cc]: How about this?

void Application::Terminate()
void Runtime::ExitApplication()
void RuntimeUIDelegateDesktop::OnApplicationExitRequested() 
void NativeAppWindowViews::ExitApplication()

**Comment by joone** [comment]
@pozdnyakov PTAL

File: runtime/browser/runtime.cc [hunk 13]
```diff
@@ -192,6 +195,15 @@ void Runtime::CloseContents(content::WebContents* source) {
     observer_->OnRuntimeClosed(this);
 }
 
+void Runtime::ExitApplication() {
```
Comment (pozdnyakov) [inline runtime/browser/runtime.cc]: maybe `RequestApplicationExit`

File: runtime/browser/runtime.cc [hunk 14]
```diff
@@ -192,6 +195,15 @@ void Runtime::CloseContents(content::WebContents* source) {
...
+  auto app_service =
+      XWalkRunner::GetInstance()->app_system()->application_service();
+  int rph_id = web_contents()->GetRenderProcessHost()->GetID();
+  application::Application* app =
+      app_service->GetApplicationByRenderHostID(rph_id);
+  app->Terminate();
```
Comment (pozdnyakov) [inline runtime/browser/runtime.cc]: would  be better just `observer_->OnApplicationExitRequested(this);` // `Application` is implementing `Runtime::Observer` interface, so further:
```
OnApplicationExitRequested(Runtime*  runtime)  {
  Terminate();
}
```

Comment (pozdnyakov) [inline runtime/browser/runtime.cc]: later we should consider renaming `Runtime::Observer` to `Runtime::Delegate` or just add `Runtime::Delegate`

**Comment by joone** [comment]
@pozdnyakov PTAL again.

**Comment by pozdnyakov** [comment]
lgtm, thanks
